### ISSUE TITLE
Encrypted match keys current site may conflict with fenced frames

### ISSUE THREAD

**Issue by dominiccooney** [patcg-individual-drafts/ipa#19]
The end-to-end doc talks about [`get_encrypted_match_key`](https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#client-side-getting-the-encrypted-match-key-report) and "the report collector (acting on behalf of a specific source/trigger website/app)" and reports containing "current_website/app."

My understanding is the "current website" means, approximately, the site displayed in the URL bar; a common use case would be a script from an ad network, running in an iframe displaying an ad, would call `get_encrypted_match_key` to generate an impression event which includes the top-level site.

There is a proposal for an iframe-like thing, [Fenced Frames,](https://github.com/WICG/fenced-frame/) with many use cases related to ads. Many fenced frames goals hinge on cutting communication between the fenced frame and the containing frame. For example, in [Fenced Frame Issue #38](https://github.com/WICG/fenced-frame/issues/38), @shivanigithub said:

> [sandboxed iframes] can still have communication with the top-level frame via window.top, window.parent, joint history, post message (I think) etc. and **fenced frames do not have any of those.**

(Emphasis mine.) Digging through the Chromium source in lieu of a more detailed spec, `blink::Frame::Parent(FrameTreeBoundary::kFenced)` does not cross the fenced frame boundary and this is what `window.top` and `location.ancestorOrigins` use for their walks. Thus it appears that Fenced Frames intend to stop the framed content from learning the top-level site identity.

This potentially puts IPA and Fenced Frames into conflict: IPA plans to share the top-level site with the framed content; Fenced Frames plans to hide it.

Here's a couple of potential paths forward:

First, Fenced Frames could change to not hide the top-level site. I'm not acutely aware of the tradeoffs fenced frames are making, and I understand their requirement to limit cross-domain communication, but it seems like a serious defect that ads cannot detect what site they are appearing on. It is reasonable and typical for advertisers to want to verify they are not displayed on adult sites.

Second, IPA could share the top-level site, but only with helpers. As written this is what `get_encrypted_match_key` suggests, but this interaction with fencedframe might be a good issue to flag for TAG. Naturally there may also be privacy budget implications if people were forced to use fencedframe widely. For example, a report collector could no longer independently discard event spam from unrelated origins. (But the helper network could.)

Note, in #5 @martinthomson suggested moving some repetitive fields, like current site, out of the encrypted report and into the HPKE associated data to make batched reports smaller. This is a good idea. However this requires the callee to know the plaintext top-level site.

**Comment by csharrison** [comment]
Thanks @dominiccooney for filing, I have it on my TODO list to think through IPA's interaction w/ Fenced Frames / FLEDGE.

> First, Fenced Frames could change to not hide the top-level site. I'm not acutely aware of the tradeoffs fenced frames are making, and I understand their requirement to limit cross-domain communication, but it seems like a serious defect that ads cannot detect what site they are appearing on. It is reasonable and typical for advertisers to want to verify they are not displayed on adult sites.

Since Fenced Frames will contain potentially cross-site user information (like your ad interest groups, etc.),  and opaque-ads fenced frames have [access to network](https://github.com/WICG/fenced-frame/blob/master/explainer/modes.md#opaque-ads), allowing a join with the top level site recovers cookie-like behavior. The brand safety aspects of creative placement typically take place prior to ads rendering in the fenced frame (e.g. in a FLEDGE [auction](https://github.com/WICG/turtledove/blob/main/FLEDGE.md#21-initiating-an-on-device-auction)).

> Second, IPA could share the top-level site, but only with helpers. As written this is what get_encrypted_match_key suggests, but this interaction with fencedframe might be a good issue to flag for TAG. Naturally there may also be privacy budget implications if people were forced to use fencedframe widely. For example, a report collector could no longer independently discard event spam from unrelated origins. (But the helper network could.)

For spam filtering there might be an opportunity here to use new crypto (e.g. something like blind signatures like [Trust Tokens](https://github.com/WICG/trust-token-api)) to bind a coarse-grained trust signal obtained in the context of the top-level site to the match key generated in the Fenced Frame.

One other issue with match keys generated in the fenced frame is that the report collector would ideally like to know some information about the context in which a match key was generated for querying (e.g. getting aggregate reports broken out by publisher). The beauty of IPA is that the encrypted match key is implicitly bound to its context which is publicly known to the collector. When this is no longer the case the query model breaks down. How could we extend IPA to make queries where a report collectors knowledge is "split" across contexts (top level site, fenced frame)?

**Comment by martinthomson** [comment]
It might be that the simplest way to address this sort of problem is to cut fenced frames off entirely.

The might be as simple as preventing the fenced frame from invoking IPA, with no way to use permissions policy to delegate the capability.  This preserves the isolation of the fenced frame, but also leaves the fenced frame without any means of accessing measurement capabilities.

If the fenced frame is able to learn the site identity (as a registrable domain), then providing it with access to `getEncryptedMatchKey()` might still work.  However, as Charlie says, combining even that much information with network access could permit a cross-site join.  I'm not 100% confident that stuff like shared storage or FLEDGE don't already let that much information out though[^1], so it's possible that you don't have to completely rule out giving out a registrable domain.  That means invoking the API as though it were in a fresh top-level context with a registrable domain matching the true top-level context.  Critically, it means doing a fresh encryption of the match key rather than using the one that you might already have stored for the true top-level site.

> For spam filtering there might be an opportunity here to use new crypto [...]

It seems like this arm of the discussion relates more to the design of fenced frames than it does IPA.

[^1] Indeed, I'm fairly confident that you could use the extra access provided to the isolated realm created for shared storage to segment responses so that one of your URLs is only returned for a chosen top-level site.  Then you could be fairly confident in the site identity when that content is loaded.  It shouldn't be hard to do that for one or two high-value sites (or 7[^2]), indeed, this might even be considered a useful or "legitimate" application of the targeting.

[^2]: However, releasing registrable domain creates a risk that bad actors use the public suffix list to create sites that act as high entropy identifiers.  Or even a small set of sites that can be used to progressively segment (and then identify) a population.  It's a little awkward having that sort of name appear in the URL bar, but I'm sure that's not a significant impediment in practice.  For something like FLEDGE or shared storage, where you have a (magical[^3]) k-anonymity service, it's possible that we could use the same magic on the registrable domain of the top-level site to ensure that enough people have visited that site - for that opaque URL - so that there is no additional information leakage.  That increases load on the k-anonymity service, but it might work[^3].

[^3]: Under the same assumptions that apply to the k-anonymity service.  I have some reservations about any claims that this sort of service can be deployed.

**Comment by benjaminsavage** [comment]
> How could we extend IPA to make queries where a report collectors knowledge is "split" across contexts (top level site, fenced frame)?

# Trigger Fanout Queries

There is actually no problem for "trigger fanout queries". Those only require you to know what website/app the ad is promoting. Presumably you would know the ad within the context of the fenced frame. If you know the ad, you'd know the domain/app it was promoting and the corresponding helper-node consortium and match-key provider. 

Assuming the fenced-frame had access to network, it could send this report to the advertiser buying the ads, who could use it in "trigger-fanout" queries. You would just be limited in terms of what sort of breakdowns are possible. For example, you couldn't do a "website/app where this ad was shown" breakdown. You could only do breakdowns based on information known at the time the ad was selected.

# Source Fanout Queries

The more challenging problem is to support "source fanout queries" for the website/app displaying the ad. Here's a sketch of one way to approach this:

1. Assume that there is a set of domains that are all using the same "report collector". Assume that they have opted to "pool" their privacy budgets together as alluded to in #10 
2. Assume that this "report collector" will only run "source-fanout" queries that measure the entire set of domains (no per-domain queries) as this is the most efficient way to make use of their now pooled privacy budged.
3. When invoking the "get_encrypted_match_key()" method from within the fenced-frame, do not send in the name of the specific app/website where the ad is being shown, instead just specify the common "report collector" that services all of these apps / websites. The calling code would provide the match-key provider and helper-node consortium that this pooled "report collector" has specified.

This should work - and avoid the need to know exactly what the hosting app/website was. You'd basically just know "This ad was shown somewhere on the network." You could still use it in source-fanout queries, (i.e. count the total number of website conversions generated by the ads shown in this 6-hour window across the network). As before, you'd be limited in the types of breakdowns you could do. Specifically - you wouldn't be able to do an app/website breakdown.

# How could we get an app/website breakdown?

Both of the sections above mention a common limitation: the inability to get an app/website breakdown. This seems like it's going to be a common ask, so how could we support it in a private way?

I think it might be possible...

In a normal IPA query, the report collector:
- Is aware of what breakdown key is associated with each event
- generates a 3-way secret sharing of the breakdown key
- encrypts each secret share towards one of the 3 helper nodes

We cannot do this here, but here's what we could do:
- The report collector's code running within the fenced frame could ask the browser/os to look up a private value it's not allowed to see (e.g. the top-level domain)
- The browser/os could secret share and encrypt the shares itself
- The report collector could receive (and transmit over the network) this encrypted_breakdown_key for use in subsequent IPA queries.

Basically, the report collector would just ask the browser to do a few of the steps for it (the secret sharing and encrypting steps).

**Comment by csharrison** [comment]
I agree with @benjaminsavage that one way to fix in IPA is to move away from the "late-binding" breakdown key generation into something more like ARA with fixed breakdown keys generated when the event occurs. However, it's worth thinking through techniques that can recover more of the flexibility of IPA.

Off the top of my head, what if we could split the match key into two pieces that are somehow internally linked together, one piece from the top-level context, and one piece from the fenced context: M_t, and M_f. The report collector would be able to learn each individual match key associated with the particular context it is associated with, but their "linkage" would be hidden to the collector.

```mermaid
sequenceDiagram
    participant T as Top level context
    participant F as Fenced frame
    participant R as Report collector
    T->>R: get_match_key() + top level context
    F->>R: get_match_key_linked_to_top() + fenced context
```

At query time, the report collector has a batch of source match keys, partitioned by type (fenced, or top), and can generate breakdown keys on-demand as usual for each type. Some new IPA functionality could:
- Internally match together the match key pairs (when they exist)
- For each pair, combine their breakdown key (e.g. concatenate the keys) and dedupe the pair into one row
- Run a query as normal on the combined breakdown key

It seems like a model like this could allow you to run on-demand queries that are as dynamic / flexible as IPA normally: 
- top level domain x campaign
- "spamminess signal" x campaign

There are lots of details to work out, but I think it's a direction worth exploring to keep all the (IMO) best aspects of IPA working here.

**Comment by martinthomson** [comment]
I can see that working in terms of new types event, each incomplete without the other.  Provided that you synchronized their collection somehow (either in time, or by adding another field under encryption), you could add data to one that is used by the other.  It increases the complexity of the attribution logic, but we've already established that those costs are not dominant.

I remain concerned that delivery of the two events might allow the report collector to connect the two, so we're back to concerns about the privacy properties of fenced frames, but this design doesn't appear to make it worse.

**Comment by csharrison** [comment]
@martinthomson here's an approach which fixes the problem you identified of linking the two reports, and works even if fenced frames have restricted network access:

1. Generate a report at interest-group join time on the advertiser site.
2. In the FLEDGE auction, emit a contextual report when the auction is finished with encrypted information linking back to the report in (1).

Essentially the reports in (1) serve to "augment" a query by providing x-site information to the source side of things. It may be worth thinking through this general capability within the context of IPA.
